### ISSUE TITLE
How to use flatMap (.Merge) within SignalProducer chain RAC 3.0

### ISSUE THREAD

**Issue by aloco** [ReactiveCocoa/ReactiveCocoa#2123]
Hi there, I just started to use ReactiveCocoa 3.0 for my network layer.

Sometimes I have to make a request which could invoke an unknown number of other requests after completion. e.g. a fetched model with multiple image URL´s, and every image should get preloaded.
What I want to achieve is a mechanism which invokes the model download and the image downloads with a single `start` and multiple `next` calls, which are receiving the images (not sure if that is possible with `SignalProducer`)

Please be aware that I could not fetch the images within one single call

My first attempt was something like this

```
fetchModel // returns SignalProducer<Model, NSError>
    |> flatMap(.Concat) { model -> SignalProducer<UIImage, NSError> in
        var signal = SignalProducer<UIImage, NSError>.empty
        for url in model.urls {
            signal |> flatMap(.Merge) { v in
                return fetchImage(url) // returns SignalProducer<UIImage, NSError>
            }
        }
        return signal
    }
    |> start(error: { error in
        println(error)
    }, next: { image in // Expecting multiple calls with a fetched UIImage
        println(image)
    }, completed: {
        println("completed")
    })

```

But this just completes without any next call and without invoking fetchImage(url)

I would be happy if anyone could lead me into the right direction, thank you ✌🏼

**Comment by neilpa** [comment]
Signal producers are value types and applying an operator to them creates a new one, it doesn't modify the existing one. So each iteration through the loop you're ignoring the returned producer from the application of `flatMap(.Merge)`. Even if they weren't value types this still wouldn't work because you're mapping over an empty signal.

Something like this should do what you want

```swift
fetchModel // returns SignalProducer<Model, NSError>
    // you can switch this to .Merge if you want to fetch in parallel instead of serially
    |> flatMap(.Concat) { model -> SignalProducer<UIImage, NSError> in
        let fetches = map(model.urls) { fetchImage($0) }
        return SignalProducer(values: fetches)
    }
```

**Comment by neilpa** [comment]
I've also written a simple [`uncollect`](https://github.com/neilpa/Rex/blob/4f35289761fe538cf996ca6e188838a408f1da1d/Source/Signal.swift#L83-L99) operator (inverse of [`collect`](https://github.com/ReactiveCocoa/ReactiveCocoa/blob/362c55c0ef976f7ceccc60e63c3f9001d3384748/ReactiveCocoa/Swift/Signal.swift#L265-L268)) for lifting an array of values into the signal. Personally, I find that a bit easier to reason about than the nested flattens approach above.

```swift
fetchModel
    |> map { model in model.urls }
    |> uncollect
    |> flatMap(.Merge) { url in fetchImage(url) }

**Comment by aloco** [comment]
ahh thanks, I was always wondering how to use `SignalProducer(values:) `
Your `uncollect` operator looks very handy for such use cases

In your first post, Is there a special reason to use `flatMap(.Latest)` instead of `flatMap(.Concat)`?  
For my understanding `flatMap(.Concat)` seems more logical since it waits for `completion` of the fetchModel Signal

**Comment by neilpa** [comment]
> In your first post, Is there a special reason to use flatMap(.Latest) instead of flatMap(.Concat)?

For the outer producer it doesn't really matter since I assume `fetchModel` is only sending a single value. In cases like that I usually default to `Latest`.

> For my understanding flatMap(.Concat) seems more logical since it waits for completion of the fetchModel Signal.

Nope, it immediately starts the first producer generated from the first mapped value. It's just that each subsequent producer is deferred until the preceding one finishes. That isn't affected by completion of the outer producer. Hope that makes sense.

**Comment by aloco** [comment]
thank you so much for clarification, that helps a lot 👍🏻 

I am just wondering if there is an easy way to define a maximum number of concurrent started `SignalProducer` within `flatten(.Merge)` e.g. to run maximum 5 concurrent requests

I played around with the `take` operator but that one acts like a filter which just filters the results of the already started `SignalProducer`

**Comment by alessandroorru** [comment]
how does this map on RAC 4 ? I'm running into a similar issue:

1) I start a SignalProducer for a network call to fetch a list of IDs of resources
2) I have to start a new SignalProducer on another endpoint for each one of the ID of the array

as an output I would like to receive the results of the calls as next events

suggestions?

**Comment by aloco** [comment]
Basically with RAC 4 this works the same way like before. Just replace ```|>```with `.`

In your solution, you could move `.flatten(.Merge)` to `SignalProducer(values: signalProducers)`, therefore you avoid `SignalProducer<SignalProducer<Resource, MyError>, MyError>`
```
client.resourceList()
    .flatMap(.Latest) { (ids) -> SignalProducer<Resource, MyError> in
        let signalProducers = ids.map { client.fetchResource($0) }
        return SignalProducer(values: signalProducers).flatten(.Merge)
    }
```
or instead you could use neilpa´s `uncollect`
```
client.resourceList()
    .uncollect
    .flatMap(.Merge) { id in client.fetchResource(id) }
```
